client/mm/exchange_adaptor.go

```diff
@@ -2638,6 +2509,443 @@ func (u *unifiedExchangeAdaptor) handleDEXNotification(n core.Notification) {
+// Use optimizeTransfers to set toDeposit and toWithdraw.
+func (u *unifiedExchangeAdaptor) inventory(assetID uint32, dexLot, cexLot uint64) (b *assetInventory) {
+	b = new(assetInventory)
+	u.balancesMtx.RLock()
+	defer u.balancesMtx.RUnlock()
+
+	dexBalance := u.dexBalance(assetID)
+	b.dexAvail = dexBalance.Available
+	b.dexPending = dexBalance.Pending
+	b.dexLocked = dexBalance.Locked
+	b.dex = dexBalance.Available + dexBalance.Locked + dexBalance.Pending
+	b.dexLots = b.dex / dexLot
+	cexBalance := u.cexBalance(assetID)
+	b.cexAvail = cexBalance.Available
+	b.cexPending = cexBalance.Pending
+	b.cexReserved = cexBalance.Reserved
+	b.cexLocked = cexBalance.Locked
+	b.cex = cexBalance.Available + cexBalance.Reserved + cexBalance.Pending
+	b.cexLots = b.cex / cexLot
+	b.total = b.dex + b.cex
+	return
+}
+
+// cexCounterRates attempts to get vwap estimates for the cex book for a
+// specified number of lots. If the book is too empty for the specified number
+// of lots, a 1-lot estimate will be attempted too.
+func (u *unifiedExchangeAdaptor) cexCounterRates(cexBuyLots, cexSellLots uint64) (dexBuyRate, dexSellRate uint64, err error) {
+	tryLots := func(b, s uint64) (uint64, uint64, bool, error) {
+		buyRate, _, filled, err := u.CEX.VWAP(u.baseID, u.quoteID, true, u.lotSize*s)
+		if err != nil {
+			return 0, 0, false, fmt.Errorf("error calculating dex buy price for quote conversion: %w", err)
+		}
+		if !filled {
+			return 0, 0, false, nil
+		}
+		sellRate, _, filled, err := u.CEX.VWAP(u.baseID, u.quoteID, false, u.lotSize*b)
+		if err != nil {
+			return 0, 0, false, fmt.Errorf("error calculating dex sell price for quote conversion: %w", err)
+		}
+		if !filled {
+			return 0, 0, false, nil
+		}
+		return buyRate, sellRate, true, nil
+	}
+	var filled bool
+	if dexBuyRate, dexSellRate, filled, err = tryLots(cexBuyLots, cexSellLots); err != nil || filled {
+		return
+	}
+	u.log.Tracef("Failed to get cex counter-rate for requested lots. Trying 1 lot estimate")
+	dexBuyRate, dexSellRate, filled, err = tryLots(1, 1)
+	if err != nil {
+		return
+	}
+	if !filled {
+		err = errors.New("cex book to empty to get a counter-rate estimate")
+	}
+	return
+}
+
+// bookingFees are the per-lot fees that have to be available before placing an
+// order.
+func (u *unifiedExchangeAdaptor) bookingFees(buyFees, sellFees *LotFees) (buyBookingFeesPerLot, sellBookingFeesPerLot uint64) {
+	buyBookingFeesPerLot = buyFees.Swap
+	// If we're redeeming on the same chain, add redemption fees.
+	if u.quoteFeeID == u.baseFeeID {
+		buyBookingFeesPerLot += buyFees.Redeem
+	}
+	// EVM assets need to reserve refund gas.
+	if u.quoteTraits.IsAccountLocker() {
+		buyBookingFeesPerLot += buyFees.Refund
+	}
+	sellBookingFeesPerLot = sellFees.Swap
+	if u.baseFeeID == u.quoteFeeID {
+		sellBookingFeesPerLot += sellFees.Redeem
+	}
+	if u.baseTraits.IsAccountLocker() {
+		sellBookingFeesPerLot += sellFees.Refund
+	}
```

Comment:
Doesn't really matter for now because only AccountLockers can have two assets with the same parent id, but only AccountLockers need to reserve for redemptions.

```suggestion
	buyBookingFeesPerLot = buyFees.Swap
	// EVM assets need to reserve refund gas.
	if u.quoteTraits.IsAccountLocker() {
		buyBookingFeesPerLot += buyFees.Refund
	}

	sellBookingFeesPerLot = sellFees.Swap
	if u.baseTraits.IsAccountLocker() {
		sellBookingFeesPerLot += sellFees.Refund
	}

	// If we're redeeming on the same chain, and both are EVM, add
	// redemption fees.
	if u.baseTraits.IsAccountLocker() && u.quoteFeeID == u.baseFeeID {
		buyBookingFeesPerLot += buyFees.Redeem
		sellBookingFeesPerLot += sellFees.Redeem
	}
```

Reply:
I thought about that too, but for this hypothetical non-account-locker asset that shares a fee asset with the counter-asset, it's probably correct anyway.